mlflow/genai/scorers/job.py

```diff
@@ -0,0 +1,304 @@
+    else:
+        original_tracking_uri = get_tracking_uri()
+        os.environ["_MLFLOW_GATEWAY_BASE_URL"] = original_tracking_uri
+
+    # Deserialize scorer
+    scorer_dict = json.loads(serialized_scorer)
+    scorer = Scorer.model_validate(scorer_dict)
+
+    tracking_store = _get_tracking_store()
+
+    if scorer.is_session_level_scorer:
+        result = _run_session_scorer(scorer, trace_ids, tracking_store, log_assessments)
+    else:
+        result = _run_single_turn_scorer_batch(scorer, trace_ids, tracking_store, log_assessments)
+
+    return asdict(result)
+
+
+def _fetch_traces_batch(
+    trace_ids: list[str],
+    tracking_store: AbstractStore,
+) -> tuple[dict[str, Trace], list[ScorerFailure]]:
+    """
+    Fetch traces in batch and return a mapping plus failures for missing traces.
+
+    Args:
+        trace_ids: List of trace IDs to fetch.
+        tracking_store: The tracking store instance.
+
+    Returns:
+        Tuple of (trace_id -> trace mapping, list of failures for missing traces).
+
+    Raises:
+        Exception: If batch_get_traces fails (e.g., DB error).
+    """
+    traces = tracking_store.batch_get_traces(trace_ids)
+    trace_map = {t.info.trace_id: t for t in traces}
+
+    # Detect missing traces
+    failures = [
+        ScorerFailure(
+            trace_id=trace_id,
+            error_code="TRACE_NOT_FOUND",
+            error_message=f"Trace not found: {trace_id}",
+        )
+        for trace_id in trace_ids
+        if trace_id not in trace_map
+    ]
+
+    return trace_map, failures
+
+
+def _run_session_scorer(
+    scorer: Any,
+    trace_ids: list[str],
+    tracking_store: AbstractStore,
+    log_assessments: bool,
+) -> InvokeScorerResult:
+    """
+    Run a session-level scorer on all traces as a conversation.
+
+    Reuses evaluate_session_level_scorers from the evaluation harness.
+
+    Args:
+        scorer: The scorer instance.
+        trace_ids: List of trace IDs (representing a session).
+        tracking_store: The tracking store instance.
+        log_assessments: Whether to log assessments to the traces.
+
+    Returns:
+        Result dictionary with success, assessments, and failures.
+    """
+    trace_map, failures = _fetch_traces_batch(trace_ids, tracking_store)
+
+    # For session scorers, we need all traces - fail if any are missing
+    if failures:
+        return InvokeScorerResult(
+            trace_ids=trace_ids,
+            success=False,
+            failures=failures,
+        )
+
+    # Preserve order of traces as requested
+    traces = [trace_map[tid] for tid in trace_ids]
+
+    session_items = [EvalItem.from_trace(t) for t in traces]
+
+    # Get session_id from the first trace's metadata
+    first_session_item = get_first_trace_in_session(session_items)
+    trace_metadata = first_session_item.trace.info.trace_metadata or {}
+    session_id = trace_metadata.get(TraceMetadataKey.TRACE_SESSION)
+
+    if not session_id:
```

Comment:
I think we should validate this for each trace in `session_items`, unless we already validate somewhere?